### PR TITLE
fix: Only store valid ip addresses in clickhouse

### PR DESCRIPTION
**Pull request by untitaker**
Related to https://github.com/getsentry/sentry/pull/14727

### CODE REVIEW

**Comment by untitaker** [comment]
Tested locally, manually.

File: snuba/processor.py [hunk 1]
```diff
@@ -105,3 +106,18 @@ def _ensure_valid_date(dt):
     if _collapse_uint32(seconds) is None:
         return None
     return dt
+
+
+def _ensure_valid_ip(ip):
```
Comment (fpacifici) [inline snuba/processor.py]: min adding typehints ? we are trying to add them consistently to the snuba code base for new code.

File: snuba/datasets/events_processor.py [hunk 2]
```diff
@@ -30,6 +30,9 @@ def extract_user(output, user):
     output['user_id'] = _unicodify(user.get('id', None))
     output['username'] = _unicodify(user.get('username', None))
     output['email'] = _unicodify(user.get('email', None))
+    # N.b.: IP address may be a hash of the real IP address (done by PII
+    # stripping). For as long as we can store arbitrary strings in Clickhouse,
+    # let's index those hashes as well.
```
Comment (fpacifici) [inline snuba/datasets/events_processor.py]: what do we gain in preserving those hashes ?
That would reduce our data integrity, so I think we should have a good reason for doing this.

Comment (untitaker) [reply snuba/datasets/events_processor.py]: If you store a hash of an IP address you might have anonymized the IP sufficiently for compliance while still being able to count distinct users

I am not sure if people do this fwiw

Comment (fpacifici) [reply snuba/datasets/events_processor.py]: I'd prefer this hash to be in another column if we keep storing it. So we do not break the assumption that the ip_address column actually contains an ip_address.

**Comment by untitaker** [comment]
I believe I have addressed all comments. Note that there is no linting or type checking happening in CI though

File: snuba/datasets/events_processor.py [hunk 3]
```diff
@@ -30,7 +31,8 @@ def extract_user(output, user):
     output['user_id'] = _unicodify(user.get('id', None))
     output['username'] = _unicodify(user.get('username', None))
     output['email'] = _unicodify(user.get('email', None))
-    output['ip_address'] = _unicodify(user.get('ip_address', None))
+    ip_addr = _ensure_valid_ip(user.get('ip_address', None))
+    output['ip_address'] = str(ip_addr) if ip_addr is not None else None
```
Comment (fpacifici) [inline snuba/datasets/events_processor.py]: > output['ip_address'] = str(ip_addr) or None

**Comment by fpacifici** [comment]
About the lint and type checking. Yes we did not add it in CI yet, because we need to ensure it would block only if you reduce the type hint coverage since the majority of the code base does not ahve type hints. We did not have time to do that yet.

**Comment by mitsuhiko** [comment]
> how has this been working so far? Did we not hash PII ?

Only relay has this feature and it has not been particularly used yet. Since it would have been rejected and nulled at ingest you could not actually get that in so far.

We also discussed just not allowing PII stripping on these fields other than to remove it entirely.

File: snuba/datasets/events_processor.py (hunk 3 above)
Comment (untitaker) [reply snuba/datasets/events_processor.py]: This would stringify a `ip_addr = None` to `"None"` which is not falsy.

Comment (fpacifici) [reply snuba/datasets/events_processor.py]: meh. ok.

**Pull request merged**